File: go/legacy.go
package iprangeformat

import (
	"bytes"
	"math/big"
	"strconv"
	"unicode/utf8"
)

// Legacy is a parsed legacy iprange binary file (v1.0 IPv4 / v2.0 IPv6), read for
// migration to v3. See .agents/sow/specs/legacy-binary-format.md. The IPv6 keys are
// already transposed to v3's hi-then-lo form.
type Legacy struct {
	IsV6      bool
	Optimized bool
	Lines     uint64
	UniqueIPs string // decimal, exactly as the header carried it
	RangesV4  [][2]Ipv4Key
	RangesV6  [][2]Ipv6Key
}

const (
	legacyMagicV4 = "iprange binary format v1.0"
	legacyMagicV6 = "iprange binary format v2.0"
)

var (
	legacyMarkerLE = [4]byte{0x4D, 0x3C, 0x2B, 0x1A} // 0x1A2B3C4D little-endian
	legacyMarkerBE = [4]byte{0x1A, 0x2B, 0x3C, 0x4D}
)

// ParseLegacy parses a legacy binary file or returns an error.
func ParseLegacy(b []byte) (*Legacy, error) {
	pos := 0
	magic, err := readLegacyLine(b, &pos)
	if err != nil {
		return nil, err
	}
	var isV6 bool
	switch magic {
	case legacyMagicV4:
		isV6 = false
	case legacyMagicV6:
		isV6 = true
	default:
		return nil, errInvalidInput("not a legacy iprange binary file")
	}

	if isV6 {
		fam, err := readLegacyLine(b, &pos)
		if err != nil {
			return nil, err
		}
		if fam != "ipv6" {
			return nil, errInvalidInput("legacy v2.0 missing ipv6 line")
		}
	}

	optLine, err := readLegacyLine(b, &pos)
	if err != nil {
		return nil, err
	}
	var optimized bool
	switch optLine {
	case "optimized":
		optimized = true
	case "non-optimized":
		optimized = false
	default:
		return nil, errInvalidInput("legacy optimized flag line malformed")
	}

	recordSize, err := parsePrefixed(b, &pos, "record size ")
	if err != nil {
		return nil, err
	}
	expectRS := uint64(8)
	if isV6 {
		expectRS = 32
	}
	if recordSize != expectRS {
		return nil, errInvalidInput("legacy record size mismatch for family")
	}
	records, err := parsePrefixed(b, &pos, "records ")
	if err != nil {
		return nil, err
	}
	bytesField, err := parsePrefixed(b, &pos, "bytes ")
	if err != nil {
		return nil, err
	}
	lines, err := parsePrefixed(b, &pos, "lines ")
	if err != nil {
		return nil, err
	}
	uniqueLine, err := readLegacyLine(b, &pos)
	if err != nil {
		return nil, err
	}
	uniqueStr, ok := stripPrefix(uniqueLine, "unique ips ")
	if !ok {
		return nil, errInvalidInput("legacy unique ips line prefix mismatch")
	}

	payload, ok1 := mul64(records, recordSize)
	payload, ok2 := add64(payload, 4)
	if !ok1 || !ok2 {
		return nil, errOverflow("legacy bytes field")
	}
	if bytesField != payload {
		return nil, errInvalidInput("legacy bytes field inconsistent")
	}
	if lines < records {
		return nil, errInvalidInput("legacy lines < records")
	}

	if pos+4 > len(b) {
		return nil, errInvalidInput("legacy truncated before marker")
	}
	var marker [4]byte
	copy(marker[:], b[pos:pos+4])
	var le bool
	switch marker {
	case legacyMarkerLE:
		le = true
	case legacyMarkerBE:
		le = false
	default:
		return nil, errInvalidInput("legacy endianness marker invalid")
	}
	pos += 4

	body := b[pos:]
	need, ok := mul64(records, recordSize)
	if !ok || uint64(len(body)) != need {
		return nil, errInvalidInput("legacy payload length mismatch / trailing data")
	}

	res := &Legacy{IsV6: isV6, Optimized: optimized, Lines: lines, UniqueIPs: uniqueStr}
	if isV6 {
		if err := parseLegacyV6(res, body, int(records), le, optimized, uniqueStr); err != nil {
			return nil, err
		}
	} else {
		if err := parseLegacyV4(res, body, int(records), le, optimized, uniqueStr); err != nil {
			return nil, err
		}
	}
	return res, nil
}

func parseLegacyV4(res *Legacy, body []byte, n int, le, optimized bool, uniqueStr string) error {
	ranges := make([][2]Ipv4Key, 0, n)
	for i := 0; i < n; i++ {
		r := body[i*8 : i*8+8]
		addr := rdU32(r[0:4], le)
		bcast := rdU32(r[4:8], le)
		if addr > bcast {
			return errInvariant("legacy record addr > broadcast")
		}
		ranges = append(ranges, [2]Ipv4Key{Ipv4Key(addr), Ipv4Key(bcast)})
	}
	unique, okp := new(big.Int).SetString(uniqueStr, 10)
	if !okp {
		return errInvalidInput("legacy unique ips not a number")
	}
	if unique.Sign() < 0 || unique.BitLen() > 64 {
		return errInvalidInput("legacy v4 unique ips exceeds u64") // matches Rust's u64 parse
	}
	if big.NewInt(int64(len(ranges))).Cmp(unique) > 0 {
		return errInvalidInput("legacy unique ips < records")
	}
	if optimized {
		sum := new(big.Int)
		for i, rg := range ranges {
			if i > 0 && rg[0] <= ranges[i-1][1] {
				return errInvariant("legacy optimized records not sorted/disjoint")
			}
			sz := big.NewInt(int64(uint64(rg[1]-rg[0]) + 1))
			sum.Add(sum, sz)
		}
		if sum.Cmp(unique) != 0 {
			return errInvalidInput("legacy unique ips != sum of ranges")
		}
	}
	res.RangesV4 = ranges
	return nil
}

func parseLegacyV6(res *Legacy, body []byte, n int, le, optimized bool, uniqueStr string) error {
	ranges := make([][2]Ipv6Key, 0, n)
	for i := 0; i < n; i++ {
		r := body[i*32 : i*32+32]
		addr := rdV6(r[0:16], le)
		bcast := rdV6(r[16:32], le)
		if addr.cmp(bcast) > 0 {
			return errInvariant("legacy record addr > broadcast")
		}
		ranges = append(ranges, [2]Ipv6Key{addr, bcast})
	}
	unique, okp := new(big.Int).SetString(uniqueStr, 10)
	if !okp {
		return errInvalidInput("legacy unique ips not a number")
	}
	if unique.Sign() < 0 || unique.BitLen() > 128 {
		return errInvalidInput("legacy v6 unique ips exceeds u128") // matches Rust's u128 parse
	}
	if big.NewInt(int64(len(ranges))).Cmp(unique) > 0 {
		return errInvalidInput("legacy unique ips < records")
	}
	if optimized {
		sum := new(big.Int)
		one := big.NewInt(1)
		for i, rg := range ranges {
			if i > 0 && rg[0].cmp(ranges[i-1][1]) <= 0 {
				return errInvariant("legacy optimized records not sorted/disjoint")
			}
			sz := new(big.Int).Sub(v6Big(rg[1]), v6Big(rg[0]))
			sz.Add(sz, one)
			sum.Add(sum, sz)
		}
		if sum.Cmp(unique) != 0 {
			return errInvalidInput("legacy unique ips != sum of ranges")
		}
	}
	res.RangesV6 = ranges
	return nil
}

// rdV6 reads a legacy 16-byte IPv6 address into a v3 hi-then-lo key (transposition).
func rdV6(b []byte, le bool) Ipv6Key {
	if le {
		// little-endian writer stores {lo, hi}: bytes 0-7 = lo, 8-15 = hi.
		return Ipv6Key{Hi: rdU64(b[8:16], true), Lo: rdU64(b[0:8], true)}
	}
	// big-endian writer stores {hi, lo}: bytes 0-7 = hi, 8-15 = lo.
	return Ipv6Key{Hi: rdU64(b[0:8], false), Lo: rdU64(b[8:16], false)}
}

func rdU32(b []byte, le bool) uint32 {
	if le {
		return uint32(b[0]) | uint32(b[1])<<8 | uint32(b[2])<<16 | uint32(b[3])<<24
	}
	return uint32(b[3]) | uint32(b[2])<<8 | uint32(b[1])<<16 | uint32(b[0])<<24
}

func rdU64(b []byte, le bool) uint64 {
	if le {
		return le64(b)
	}
	return be64(b)
}

func le64(b []byte) uint64 {
	return uint64(b[0]) | uint64(b[1])<<8 | uint64(b[2])<<16 | uint64(b[3])<<24 |
		uint64(b[4])<<32 | uint64(b[5])<<40 | uint64(b[6])<<48 | uint64(b[7])<<56
}
func be64(b []byte) uint64 {
	return uint64(b[7]) | uint64(b[6])<<8 | uint64(b[5])<<16 | uint64(b[4])<<24 |
		uint64(b[3])<<32 | uint64(b[2])<<40 | uint64(b[1])<<48 | uint64(b[0])<<56
}

func v6Big(k Ipv6Key) *big.Int {
	x := new(big.Int).SetUint64(k.Hi)
	x.Lsh(x, 64)
	return x.Or(x, new(big.Int).SetUint64(k.Lo))
}

func readLegacyLine(b []byte, pos *int) (string, error) {
	start := *pos
	if start >= len(b) {
		return "", errInvalidInput("legacy header truncated")
	}
	nl := bytes.IndexByte(b[start:], '\n')
	if nl < 0 {
		return "", errInvalidInput("legacy header line missing newline")
	}
	line := b[start : start+nl]
	if !utf8.Valid(line) {
		return "", errInvalidInput("legacy header line not UTF-8")
	}
	*pos = start + nl + 1
	return string(line), nil
}

func stripPrefix(line, prefix string) (string, bool) {
	if len(line) < len(prefix) || line[:len(prefix)] != prefix {
		return "", false
	}
	return line[len(prefix):], true
}

func parsePrefixed(b []byte, pos *int, prefix string) (uint64, error) {
	line, err := readLegacyLine(b, pos)
	if err != nil {
		return 0, err
	}
	rest, ok := stripPrefix(line, prefix)
	if !ok {
		return 0, errInvalidInput("legacy header line prefix mismatch")
	}
	v, err := strconv.ParseUint(rest, 10, 64)
	if err != nil {
		return 0, errInvalidInput("legacy header numeric field malformed")
	}
	return v, nil
}

func mul64(a, b uint64) (uint64, bool) {
	if a != 0 && b > maxUint64/a {
		return 0, false
	}
	return a * b, true
}

File: rust/iprange-format/src/legacy.rs
//! Read-only parser for the **legacy** iprange binary format (v1.0 IPv4 / v2.0 IPv6)
//! produced by `iprange --print-binary`, for migration to v3.
//!
//! The byte layout is specified in `.agents/sow/specs/legacy-binary-format.md`
//! (verified against real `iprange --print-binary` artifacts). The one subtlety: a
//! legacy IPv6 address is stored `lo`-then-`hi` on a little-endian writer — the
//! opposite of v3's `hi`-then-`lo` key — so migration transposes the halves.
//!
//! Requires the `alloc` feature.

extern crate alloc;
use alloc::vec::Vec;

use crate::error::{Error, Result};
use crate::key::{Ipv4Key, Ipv6Key};

const MAGIC_V4: &str = "iprange binary format v1.0";
const MAGIC_V6: &str = "iprange binary format v2.0";
const MARKER_LE: [u8; 4] = [0x4D, 0x3C, 0x2B, 0x1A]; // 0x1A2B3C4D written little-endian

/// A parsed legacy file: the IP family with its inclusive `[start, end]` ranges plus
/// the header metadata. Ranges are in file order (sorted+disjoint when `optimized`).
#[derive(Clone, Debug)]
pub enum Legacy {
    /// IPv4 (v1.0).
    V4 {
        /// Whether the file declared itself optimized (sorted + disjoint).
        optimized: bool,
        /// Header `unique ips` (total addresses covered).
        unique_ips: u64,
        /// Header `lines` (source input lines).
        lines: u64,
        /// The ranges.
        ranges: Vec<(Ipv4Key, Ipv4Key)>,
    },
    /// IPv6 (v2.0).
    V6 {
        /// Whether the file declared itself optimized.
        optimized: bool,
        /// Header `unique ips` as a 128-bit count.
        unique_ips: u128,
        /// Header `lines`.
        lines: u64,
        /// The ranges (keys already transposed to v3's hi-then-lo form).
        ranges: Vec<(Ipv6Key, Ipv6Key)>,
    },
}

/// Parse a legacy binary file. Returns an error if it is not a recognized legacy
/// file or violates the legacy loader's structural rules.
pub fn parse(bytes: &[u8]) -> Result<Legacy> {
    let mut pos = 0usize;
    let magic = read_line(bytes, &mut pos)?;
    let is_v6 = if magic == MAGIC_V4 {
        false
    } else if magic == MAGIC_V6 {
        true
    } else {
        return Err(Error::InvalidInput("not a legacy iprange binary file"));
    };

    if is_v6 {
        let fam = read_line(bytes, &mut pos)?;
        if fam != "ipv6" {
            return Err(Error::InvalidInput("legacy v2.0 missing ipv6 line"));
        }
    }

    let opt_line = read_line(bytes, &mut pos)?;
    let optimized = match opt_line {
        "optimized" => true,
        "non-optimized" => false,
        _ => return Err(Error::InvalidInput("legacy optimized flag line malformed")),
    };

    let record_size = parse_prefixed(read_line(bytes, &mut pos)?, "record size ")?;
    let expect_rs: u64 = if is_v6 { 32 } else { 8 };
    if record_size != expect_rs {
        return Err(Error::InvalidInput("legacy record size mismatch for family"));
    }
    let records = parse_prefixed(read_line(bytes, &mut pos)?, "records ")?;
    let bytes_field = parse_prefixed(read_line(bytes, &mut pos)?, "bytes ")?;
    let lines = parse_prefixed(read_line(bytes, &mut pos)?, "lines ")?;
    let unique_line = read_line(bytes, &mut pos)?;
    let unique_str = strip_prefix(unique_line, "unique ips ")?;

    // payload bytes = records*record_size + 4 (the marker).
    let payload = records
        .checked_mul(record_size)
        .and_then(|p| p.checked_add(4))
        .ok_or(Error::Overflow("legacy bytes field"))?;
    if bytes_field != payload {
        return Err(Error::InvalidInput("legacy bytes field inconsistent"));
    }
    if lines < records {
        return Err(Error::InvalidInput("legacy lines < records"));
    }

    // endianness marker. Only little-endian is accepted: the legacy C tool refuses
    // cross-endian files, and §14 of the v3 spec rejects a big-endian marker. Real
    // legacy files come from x86-64 (little-endian); we never need the BE path.
    let marker = bytes.get(pos..pos + 4).ok_or(Error::InvalidInput("legacy truncated before marker"))?;
    if marker != MARKER_LE {
        return Err(Error::InvalidInput(
            "legacy file is not little-endian (big-endian rejected, matching the C tool and §14)",
        ));
    }
    pos += 4;

    // records must be exactly the remaining bytes (no trailing data). Compute in u64
    // so `records` is never truncated by a 32-bit `usize` before the multiply.
    let body = &bytes[pos..];
    let need = records
        .checked_mul(record_size)
        .ok_or(Error::Overflow("legacy payload size"))?;
    if body.len() as u64 != need {
        return Err(Error::InvalidInput("legacy payload length mismatch / trailing data"));
    }

    if is_v6 {
        let unique_ips: u128 = unique_str.parse().map_err(|_| Error::InvalidInput("legacy unique ips not a u128"))?;
        let ranges = parse_v6_records(body, records as usize)?;
        validate_v6(&ranges, optimized, unique_ips)?;
        Ok(Legacy::V6 { optimized, unique_ips, lines, ranges })
    } else {
        let unique_ips: u64 = unique_str.parse().map_err(|_| Error::InvalidInput("legacy unique ips not a u64"))?;
        let ranges = parse_v4_records(body, records as usize)?;
        validate_v4(&ranges, optimized, unique_ips)?;
        Ok(Legacy::V4 { optimized, unique_ips, lines, ranges })
    }
}

fn parse_v4_records(body: &[u8], n: usize) -> Result<Vec<(Ipv4Key, Ipv4Key)>> {
    let mut out = Vec::with_capacity(n);
    for i in 0..n {
        let b = &body[i * 8..i * 8 + 8];
        let addr = u32::from_le_bytes([b[0], b[1], b[2], b[3]]);
        let bcast = u32::from_le_bytes([b[4], b[5], b[6], b[7]]);
        if addr > bcast {
            return Err(Error::Invariant("legacy record addr > broadcast"));
        }
        out.push((Ipv4Key(addr), Ipv4Key(bcast)));
    }
    Ok(out)
}

fn parse_v6_records(body: &[u8], n: usize) -> Result<Vec<(Ipv6Key, Ipv6Key)>> {
    let mut out = Vec::with_capacity(n);
    for i in 0..n {
        let b = &body[i * 32..i * 32 + 32];
        let addr = rd_v6(&b[0..16]);
        let bcast = rd_v6(&b[16..32]);
        if addr > bcast {
            return Err(Error::Invariant("legacy record addr > broadcast"));
        }
        // a full-IPv6-space range (size 2^128) is unrepresentable in v3 — reject it
        // here so both languages fail at the legacy layer, not at migration.
        if addr.hi == 0 && addr.lo == 0 && bcast.hi == u64::MAX && bcast.lo == u64::MAX {
            return Err(Error::InvalidInput("legacy range covers the entire IPv6 space"));
        }
        out.push((addr, bcast));
    }
    Ok(out)
}

/// Read a legacy 16-byte little-endian IPv6 address into a v3 `hi`-then-`lo` key. The
/// legacy little-endian layout stores `{lo, hi}` (bytes 0–7 = `lo`, 8–15 = `hi`), the
/// opposite of v3's key, so this transposes the halves.
fn rd_v6(b: &[u8]) -> Ipv6Key {
    Ipv6Key { hi: u64_le(&b[8..16]), lo: u64_le(&b[0..8]) }
}

fn u64_le(b: &[u8]) -> u64 {
    let mut a = [0u8; 8];
    a.copy_from_slice(&b[0..8]);
    u64::from_le_bytes(a)
}

fn validate_v4(ranges: &[(Ipv4Key, Ipv4Key)], optimized: bool, unique_ips: u64) -> Result<()> {
    if (unique_ips as u128) < ranges.len() as u128 {
        return Err(Error::InvalidInput("legacy unique ips < records"));
    }
    if optimized {
        let mut sum: u128 = 0;
        for w in 0..ranges.len() {
            let (s, e) = ranges[w];
            if w > 0 && s.0 <= ranges[w - 1].1 .0 {
                return Err(Error::Invariant("legacy optimized records not sorted/disjoint"));
            }
            sum += u128::from(e.0 - s.0) + 1;
        }
        if sum != u128::from(unique_ips) {
            return Err(Error::InvalidInput("legacy unique ips != sum of ranges"));
        }
    }
    Ok(())
}

fn validate_v6(ranges: &[(Ipv6Key, Ipv6Key)], optimized: bool, unique_ips: u128) -> Result<()> {
    if unique_ips < ranges.len() as u128 {
        return Err(Error::InvalidInput("legacy unique ips < records"));
    }
    if optimized {
        let mut sum: u128 = 0;
        for w in 0..ranges.len() {
            let (s, e) = ranges[w];
            if w > 0 && s <= ranges[w - 1].1 {
                return Err(Error::Invariant("legacy optimized records not sorted/disjoint"));
            }
            // size = e - s + 1, checked. e >= s holds (per-record check), so the
            // subtraction cannot underflow; only the full IPv6 space overflows +1.
            let size = (e.to_u128() - s.to_u128())
                .checked_add(1)
                .ok_or(Error::InvalidInput("legacy range covers the entire IPv6 space"))?;
            sum = sum.checked_add(size).ok_or(Error::Overflow("legacy unique ips sum"))?;
        }
        if sum != unique_ips {
            return Err(Error::InvalidInput("legacy unique ips != sum of ranges"));
        }
    }
    Ok(())
}

/// Read one `\n`-terminated ASCII line (returned without the `\n`), advancing `pos`.
fn read_line<'a>(bytes: &'a [u8], pos: &mut usize) -> Result<&'a str> {
    let start = *pos;
    let rel = bytes[start..]
        .iter()
        .position(|&b| b == b'\n')
        .ok_or(Error::InvalidInput("legacy header line missing newline"))?;
    let line = &bytes[start..start + rel];
    *pos = start + rel + 1;
    core::str::from_utf8(line).map_err(|_| Error::InvalidInput("legacy header line not UTF-8"))
}

fn strip_prefix<'a>(line: &'a str, prefix: &str) -> Result<&'a str> {
    line.strip_prefix(prefix)
        .ok_or(Error::InvalidInput("legacy header line prefix mismatch"))
}

fn parse_prefixed(line: &str, prefix: &str) -> Result<u64> {
    strip_prefix(line, prefix)?
        .parse()
        .map_err(|_| Error::InvalidInput("legacy header numeric field malformed"))
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Build a minimal legacy v2.0 file with a single record [addr16 || bcast16].
    fn legacy_v6(unique: &str, addr: [u8; 16], bcast: [u8; 16]) -> Vec<u8> {
        let mut b = Vec::new();
        b.extend_from_slice(b"iprange binary format v2.0\nipv6\noptimized\nrecord size 32\nrecords 1\nbytes 36\nlines 1\nunique ips ");
        b.extend_from_slice(unique.as_bytes());
        b.push(b'\n');
        b.extend_from_slice(&MARKER_LE);
        b.extend_from_slice(&addr);
        b.extend_from_slice(&bcast);
        b
    }

    #[test]
    fn rejects_full_ipv6_space() {
        // full space [::, ffff:..:ffff]: legacy saturates the count to 2^128-1.
        let bytes = legacy_v6(
            "340282366920938463463374607431768211455",
            [0u8; 16],
            [0xFFu8; 16],
        );
        assert!(matches!(parse(&bytes), Err(Error::InvalidInput(_))));
    }

    #[test]
    fn rejects_big_endian_marker() {
        let mut bytes = legacy_v6("256", [0u8; 16], {
            let mut e = [0u8; 16];
            e[0] = 0xff; // lo low byte -> a tiny non-zero range end
            e
        });
        let marker_off = bytes.len() - 32 - 4;
        bytes[marker_off..marker_off + 4].copy_from_slice(&[0x1A, 0x2B, 0x3C, 0x4D]); // BE marker
        assert!(matches!(parse(&bytes), Err(Error::InvalidInput(_))));
    }
}
